internal/stolonctl/pkg/cluster/upgrade.go

```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
+			} else {
+				for _, keeperState := range res.Spec.ClusterData.KeepersState {
+					if pod.PodIP == keeperState.ListenAddress {
+						standbyNodeNames[keeperState.ID] = pod.HostIP
+					}
+				}
+			}
 		}
-		if err = backupPostgres(config); err != nil {
-			return trace.Wrap(err)
+
+		log.Infof("master: %s, standbys: %v\n", masterNodeName, standbyNodeNames)
+		if masterNodeName == "" {
+			return trace.BadParameter("IP address of keeper master pod is empty")
 		}
-		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseBackupPostgres, crd.StolonUpgradeStatusCompleted)
+		err := upgradeControl.upgradePostgresSchema(ctx, pgMaster, masterNodeName, "")
 		if err != nil {
 			return trace.Wrap(err)
 		}
+		for id, standbyNodeName := range standbyNodeNames {
+			err = upgradeControl.upgradePostgresSchema(ctx, pgStandby, standbyNodeName, id)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+		return nil
+
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if _, err := upgradeControl.crdClient.CompleteUpgrade(res); err != nil {
+		return trace.Wrap(err)
 	}
 	return nil
 }
 
-func backupPostgres(config Config) error {
-	if err := createPgPassFile(config); err != nil {
+func (u *upgradeControl) executePhase(res *crd.StolonUpgradeResource, phase string, fn func() error) (*crd.StolonUpgradeResource, error) {
+	var err error
+	if !u.crdClient.IsPhaseCompleted(res, phase) {
+		res, err = u.crdClient.MarkPhase(res, phase, crd.StolonUpgradeStatusInProgress)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if err := fn(); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		res, err = u.crdClient.MarkPhase(res, phase, crd.StolonUpgradeStatusCompleted)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return res, nil
+}
+
+func (u *upgradeControl) checkStatus(res *crd.StolonUpgradeResource, phase string) (*crd.StolonUpgradeResource, error) {
+	var err error
+	if !u.crdClient.IsPhaseCompleted(res, phase) {
+		res, err = u.crdClient.MarkPhase(res, phase, crd.StolonUpgradeStatusInProgress)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		status, err := GetStatus(u.config)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		masterStatus, err := status.getMasterStatus()
+		if err != nil {
+			return nil, trace.Wrap(err, "cannot start upgrade")
+		}
+
+		if !masterStatus.Healthy {
+			return nil, trace.BadParameter("cannot start upgrade, keeper master %s is unhealthy.",
+				masterStatus.PodName)
+		}
+		log.Infof("Keeper master: pod=%s, healthy=%v, host=%s", masterStatus.PodName,
+			masterStatus.Healthy, masterStatus.HostIP)
+
+		res, err = u.crdClient.UpdateClusterInfo(res, *status.ClusterData, status.PodsStatus)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		res, err = u.crdClient.MarkPhase(res, phase, crd.StolonUpgradeStatusCompleted)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return res, nil
+}
+
+func (u *upgradeControl) backupPostgres() error {
```

Comment:
One item I'm a little curious about. I didn't track this all the way through, but it looks like the default for BackupPath is /stolon/backup.sql, but I think the container mounts /stolon-data. So I would double check that the backup is being stored persistently. 

Also, I suspect restore could be a bit tricky, if the pod get's scheduled to a node that isn't the one that took the backup.

Reply:
Oh, that's a good catch.